### PR TITLE
Adding ability to update hashes

### PR DESCRIPTION
**Pull request by asewnath**
## Description

Addresses discussion in https://github.com/GEOS-ESM/swell/issues/345.

This PR adds a tool to automatically update repositories with pinned versions based on an input date. Example of tool's interface:

```jedi_bundle_update_hash 2024-7-29```

This executable takes in a date with `%Y-%m-%d` format, retrieves commit histories for each repository/default_branch in `pinned-versions.yaml`, and then returns the commit hash for _the latest commit before the given date_. It will then update `pinned-versions.yaml` in the package's core directory with the updated hashes.

In order to see these pinned versions in the `build.yaml` file, add the pinned version flag:
```jedi_bundle -pv```. Then proceed as normal with `Clone`, `Configure`, and `Make`.

### Dependencies
- In order to test this PR completely, #41 needs to be merged.

### CODE REVIEW

**Comment by asewnath** [comment]
@Dooruk take a look at the `pinned_versions.yaml` file in this PR and let me know which repos need to be added to the list and which ones don't need to be in the list (because we only point to one stable version instead of a commit, etc.)

**Comment by Dooruk** [comment]
> @Dooruk take a look at the `pinned_versions.yaml` file in this PR and let me know which repos need to be added to the list and which ones don't need to be in the list (because we only point to one stable version instead of a commit, etc.)

Ones needed to be pinned in addition to the existing list:
- soca
- iodaconv
- geos-aero
- gsw
- ufo-data

These following are required packages but they point to certain branches. We still need the hashes though:
- icepack
- fv3

~Optional ones needed to pinned~ These have some external permission requirements, we are skipping them for now.
~- ropp-ufo~
~- rttov~

External optional one requiring pin(coming with https://github.com/GEOS-ESM/jedi_bundle/pull/41):
- da-utils ([source](https://github.com/noaa-emc/da-utils))

Ones have to be pinned until #41 merges (I will take them out afterwards):
- fms
- mom6

Ones not needed to be pinned (because we use tags for them or load them as modules):
- gsibec
- crtm

thanks @asewnath.

**Comment by asewnath** [comment]
Awesome, thanks @Dooruk

**Review by Dooruk — COMMENTED** [review]
This works like a charm and I compared against `git rev-list -1 --before=2024-06-18 develop` command, hashes match.

This works for JEDI builds prior to the FMS update, so we could merge this one first while we identify impacts of recent changes on Ricardo's results.

Is there a reason you added `update_hash` as a second entry point rather than `jedi_bundle update_hash`?

Could you update the documentation how to use this with a simple example?

**Comment by asewnath** [comment]
> This works for JEDI builds prior to the FMS update, so we could merge this one first while we identify impacts of recent changes on Ricardo's results.

In order to merge this PR first, I'll probably change the build order for `saber` and `vader` so I can be sure that at least `Configure` is working properly with the list of pinned hashes.

> Is there a reason you added `update_hash` as a second entry point rather than `jedi_bundle update_hash`?

I wouldn't consider `update_hash` an entry point, it's more like the implentation.  I wrote this first before deciding how a user was going to interface with it. Using the `jedi_bundle_update_hash` executable was the client interface I ultimately decided on. I technically don't have to have `update_hash` as a separate utility, but I think it makes the code cleaner.

> Could you update the documentation how to use this with a simple example?

Yes, of course

**Comment by Dooruk** [comment]
> In order to merge this PR first, I'll probably change the build order for `saber` and `vader` so I can be sure that at least `Configure` is working properly with the list of pinned hashes.

I tested `clone` and `configure` with June 18th (what Ricardo has) hash versions and this branch, and it worked. It works because it predates the following VADER and FMS changes:

https://github.com/orgs/JCSDA-internal/discussions/139
https://github.com/orgs/JCSDA-internal/discussions/143#discussioncomment-10075970

> I wouldn't consider `update_hash` an entry point, it's more like the implentation. I wrote this first before deciding how a user was going to interface with it. Using the `jedi_bundle_update_hash` executable was the client interface I ultimately decided on. I technically don't have to have `update_hash` as a separate utility, but I think it makes the code cleaner.

ok, sounds good!

**Comment by asewnath** [comment]
@Dooruk ah got it! I'll hold off on that then and just get some documentation written

**Comment by Dooruk** [comment]
Eh, I wrote the message once `soca` succeeded but `configure` failed on the `fv3` step. Looking at it now, there might be a difference with the hashes.

It might be going to the main repo since fv3 is a GFDL fork.

**Comment by asewnath** [comment]
> Eh, I wrote the message once `soca` succeeded but `configure` failed on the `fv3` step. Looking at it now, there might be a difference with the hashes.
> 
> It might be going to the main repo since fv3 is a GFDL fork.

`update hash` uses the default branch indicated in `build_order.yaml` for each repository. So for `fv3` it is using the following:

```
- fv3:
    repo_url_name: GFDL_atmos_cubed_sphere
    default_branch: release-stable
```
The script looks at commits only associated with the default branch. The code also uses the `repo_url_name` if specified.

**Comment by Dooruk** [comment]
You are right, the commit hashes match between `pinned_versions.yaml` and what I see with `git log`.

**Comment by Dooruk** [comment]
@rtodling tested a JEDI build I created on July 30th and the results are identical with the previous version. So now we can incorporate the FMS changes that I merged in #41  and test the hashes for July time period.

**Comment by asewnath** [comment]
`jedi_bundle Clone` and `jedi_bundle Configure` now work with pinned versions

File: src/jedi_bundle/bin/jedi_bundle_update_hash.py
Comment (ashiklom) [inline src/jedi_bundle/bin/jedi_bundle_update_hash.py]: This is _probably_ fine, but it does make me a little bit nervous. The default behavior of `.astimezone()` is to convert everything to _local_ time, which means that the same code will produce subtly different times depending on exactly _where_ it runs.

If we want to have somewhat more predictable behavior, I might do something like this to always force midnight UTC time:

```python
from datetime import datetime, date, time, UTC
# [...]
dt_object = datetime.combine(date.fromisoformat(date), time(0, 0, tzinfo=UTC))
```

File: src/jedi_bundle/config/bundles/build-order.yaml
Comment (ashiklom) [inline src/jedi_bundle/config/bundles/build-order.yaml]: Commented code should be removed, unless it has a good reason for sticking around, in which case it should be commented.

```suggestion
```

File: src/jedi_bundle/utils/update_hash.py
Comment (ashiklom) [inline src/jedi_bundle/utils/update_hash.py]: I appreciate the implementation of the binary search algorithm, but I think you can do this much more simply by taking advantage of the [GitHub API](https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28), which as an `until` parameter:

```python
commit_history_url = url + f'{name}/commits?per_page={commit_history_length}&sha={commit_pointer}&until={date.isoformat()}'
```

The last commit returned by that query will be the last commit before the indicated date.

File: src/jedi_bundle/utils/update_hash.py
Comment (ashiklom) [inline src/jedi_bundle/utils/update_hash.py]: Requests offers a cleaner syntax for query parameters:

```suggestion
                    commit_history_url = url + f'{name}/commits'                   
                    response = requests.get(commit_history_url, headers=auth_header, params={
                        "per_page": commit_history_length,
                        "sha": commit_pointer,
                        "until": date.isoformat()
                    })
```

(Also, see comment below about `until` param).

File: src/jedi_bundle/utils/update_hash.py
Comment (asewnath) [reply src/jedi_bundle/utils/update_hash.py]: 🥲

**Review by ashiklom — APPROVED** [review]

**Pull request merged**
